pkg/controller/noderef/noderef_controller.go

```diff
@@ -0,0 +1,183 @@
+// Add creates a new NodeRef Controller and adds it to the Manager with default RBAC. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileNodeRef{Client: mgr.GetClient(), scheme: mgr.GetScheme()}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("noderef-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to Machines.
+	err = c.Watch(&source.Kind{Type: &clusterv1alpha1.Machine{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileNodeRef{}
+
+// ReconcileNodeRef reconciles a Machine object.
+type ReconcileNodeRef struct {
+	client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile watches Machines.
+func (r *ReconcileNodeRef) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	// Fetch the NodeRef instance
+	machine := &clusterv1alpha1.Machine{}
+	err := r.Get(context.TODO(), request.NamespacedName, machine)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Check that the Machine hasn't been deleted or in the process.
+	if !machine.DeletionTimestamp.IsZero() {
+		return reconcile.Result{}, nil
+	}
+
+	// Check that the Machine has a cluster label.
+	if machine.Labels[clusterv1alpha1.MachineClusterLabelName] == "" {
+		klog.Warningf("Machine %q in namespace %q doesn't specify %q label, cannot set nodeRef", machine.Name, machine.Namespace, clusterv1alpha1.MachineClusterLabelName)
+		return reconcile.Result{}, nil
+	}
+
+	// Check that the Machine has a valid ProviderID.
+	if machine.Spec.ProviderID == nil || *machine.Spec.ProviderID == "" {
+		klog.Warningf("Machine %q in namespace %q doesn't have a valid ProviderID, cannot set nodeRef", machine.Name, machine.Namespace)
+		return reconcile.Result{}, nil
+	}
+
+	cluster, err := r.getCluster(context.TODO(), machine)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	clusterClient, err := remote.NewClusterClient(r.Client, cluster)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	corev1Client, err := clusterClient.CoreV1()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Get the Node reference.
+	nodeRef, err := r.getNodeReference(corev1Client, machine)
+	if err != nil {
+		if err == ErrNodeNotFound {
+			return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Update Machine.
+	machine.Status.NodeRef = nodeRef
+	if err := r.Client.Update(context.TODO(), machine); err != nil {
```

Comment:
UpdateStatus?

Reply:
The controller-runtime Client actually only has Update

Reply:
Just checked - looks like it's `r.Client.Status().Update(...)`